src/generator.cs

```diff
@@ -3699,6 +3699,9 @@ void GetReturnsWrappers (MethodInfo mi, MemberInformation minfo, Type declaringT
 				print ("IntPtr retvalarrtmp;");
 				cast_a = "((retvalarrtmp = ";
 				cast_b = ") == IntPtr.Zero ? null : (";
+				if (bindAsT == null) {
+					throw new BindingException (1067, true, "BindAs attribute of selector \"{0}\" is missing '[]' in the definition", minfo.selector);
```

Comment:
I think we can do a lot better in the message that we throw and get a more decent information out of it. This exception is going to be misleading when we have a BindAs attribute that has been defined and that returns as null due to a type mismatch.

I would look at the GetBindAsAttribute method. There we are doing:

```csharp
BindAsAttribute rv;
		if (cu != null && (rv = AttributeManager.GetCustomAttribute<BindAsAttribute> (cu)) != null)
			return rv;

		var minfo = cu as MethodInfo;
		if (minfo?.ReturnParameter != null && (rv = AttributeManager.GetCustomAttribute<BindAsAttribute> (minfo.ReturnParameter)) != null)
			return rv;

		return null;
```
I first, would move any logic that does validation of the BindAs type there, meaning that if we have other cases in which we have a mismatch, this would work. Secondly, in the exception, we know that we tried to get the type of the BindAs attribute when the return type of a method being an Array, so at the very least, rather than saying:

> BindAs attribute of selector \"{0}\" is missing '[]' in the definition

We should be telling the developer something of the kind:

> {selector} return type does not match the type defined in the BindAs attribute, are you missing a []?

That will be a lot more helpful to developers since its is stating the problem and is possible since we have all the required information.

Reply:
As we discussed offline, there was a misinterpretation of what the error message was saying so I've left it the same. Instead of doing a null check with the result of `GetBindAsAttribute`, I've updated it so it explicitly checks for `isArray` and avoids misleading exceptions. Adding validation to the `GetBindAsAttribute` would've required significantly changing how the generator is structured.